Planning: enhance several corner case in path_bounds_decider

xmyqsh opened the pull request.

jmtao commented on modules/planning/tasks/deciders/path_bounds_decider/path_bounds_decider.cc:
```diff
@@ -392,15 +392,16 @@ Status PathBoundsDecider::GeneratePullOverPathBound(
-    std::get<2>((*path_bound)[path_bound->size() - 1 - idx]) =
-        std::get<2>((*path_bound)[std::get<3>(pull_over_configuration)]);
+  for (int idx = curr_idx + 1;
+       idx < std::min(static_cast<int>(path_bound->size()),
```
comment same as above

xmyqsh replied:
Done.

jmtao commented on modules/planning/tasks/deciders/path_bounds_decider/path_bounds_decider.cc:
```diff
@@ -392,15 +392,16 @@ Status PathBoundsDecider::GeneratePullOverPathBound(
+                      curr_idx + kNumExtraTailBoundPoint + 1); ++idx) {
+    std::get<1>((*path_bound)[idx]) = std::get<1>((*path_bound)[curr_idx]);
+    std::get<2>((*path_bound)[idx]) = std::get<2>((*path_bound)[curr_idx]);
   }
```
you are not really changing anything, right?

xmyqsh replied:
@jmtao 
Here, I have fixed up the corner case when `path_bound->size() < kNumExtraTailBoundPoint` for runtime error and when `curr_idx + kNumExtraTailBoundPoint > path_bound->size()` for logic error.

panjiacheng commented on modules/planning/tasks/deciders/path_bounds_decider/path_bounds_decider.cc:
```diff
@@ -559,9 +560,10 @@ bool PathBoundsDecider::SearchPullOverPosition(
-      const auto& pull_over_point = path_bound[static_cast<size_t>(
-          back_clear_to_total_length_ratio * static_cast<double>(i) +
-          (1.0 - back_clear_to_total_length_ratio) * static_cast<double>(j))];
+      auto idx = static_cast<size_t>(
```
Please try to give it a more meaningful name, such as "pull_over_idx", etc.

xmyqsh replied:
Done.

xmyqsh commented on modules/planning/tasks/deciders/path_bounds_decider/path_bounds_decider.cc:
```diff
@@ -351,11 +351,11 @@ Status PathBoundsDecider::GeneratePullOverPathBound(
-        std::get<2>((*path_bound)[path_bound->size() - 1 - idx]) =
-            std::get<2>((*path_bound)[curr_idx]);
+      for (int idx = curr_idx + 1;
+           idx < std::min(static_cast<int>(path_bound->size()),
```
Done.

xmyqsh commented:
@jmtao @panjiacheng 
Two main things have been fixed in this PR.
First, I have fixed up the corner case when path_bound->size() < kNumExtraTailBoundPoint for runtime error and when curr_idx + kNumExtraTailBoundPoint > path_bound->size() for logic error.

Second, I have fixed and unified the `pull_over_idx ` calculation from `(i + j) / 2` to `static_cast<size_t>(
          back_clear_to_total_length_ratio * static_cast<double>(i) +
          (1.0 - back_clear_to_total_length_ratio) * static_cast<double>(j))`.
